### PR TITLE
Implement Meta Set functionality

### PR DESCRIPTION
**Pull request by mrattle**
### TL;DR - What are you trying to accomplish?

This PR introduces the implementation of the `meta_set` method, which uses the meta parser added in https://github.com/Shopify/async-memcached/pull/66, and further builds off of the implementation changes introduced with the `meta_get` method from https://github.com/Shopify/async-memcached/pull/67. Again, there are also further small fixes to parsing behaviour that became necessary as edge cases were revealed as extended functionality was introduced and more parsing edge cases were revealed

### Details - How are you making this change?  What are the effects of this change?

Meta Set functionality [covers multi-modal set and add operations](https://github.com/memcached/memcached/blob/5609673ed29db98a377749fab469fe80777de8fd/doc/protocol.txt#L688), as well as those that use CAS semantics for updating and invalidation.  This PR covers that whole range of operations.

The bulk of the new code in this PR is in the test cases since there are a ton of different functionalities that open up with the meta set method.

===================

Benchmark results - key "foo" value "bar" and no meta flags:

```
     Running benches/bench.rs (target/release/deps/bench-62483f1fc831355c)
Gnuplot not found, using plotters backend
set_with_small_string   time:   [24.479 µs 24.563 µs 24.681 µs]
                        change: [-1.4803% -0.6244% +0.2410%] (p = 0.16 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  8 (8.00%) high severe

meta_set_with_small_key_and_value
                        time:   [23.067 µs 23.143 µs 23.281 µs]
                        change: [-1.1621% -0.4409% +0.2398%] (p = 0.22 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  3 (3.00%) high mild
  7 (7.00%) high severe

meta_set_concat_with_small_key_and_value
                        time:   [23.080 µs 23.127 µs 23.194 µs]
                        change: [-0.1273% +0.6446% +1.4933%] (p = 0.13 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low severe
  2 (2.00%) high mild
  7 (7.00%) high severe
```

Key of max length (250B), value of max size (~1MB) and no meta flags:

```
ascii set with large key and value   time:   [101.92 µs 102.16 µs 102.37 µs]
                        change: [-0.9473% -0.3876% +0.1638%] (p = 0.18 > 0.05)
                        No change in performance detected.
Found 18 outliers among 100 measurements (18.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  6 (6.00%) high mild
  8 (8.00%) high severe

bench meta_set with large key and large value
                        time:   [99.511 µs 100.62 µs 102.51 µs]
                        change: [-0.6650% +0.1177% +1.0925%] (p = 0.81 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  1 (1.00%) high mild
  6 (6.00%) high severe

bench meta_set_concat with large key and large value
                        time:   [99.723 µs 101.56 µs 104.18 µs]
                        change: [-0.8870% +0.3866% +1.8127%] (p = 0.60 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  2 (2.00%) high mild
  8 (8.00%) high severe
```

### CODE REVIEW

File: src/parser/meta.rs [hunk 1]
```diff
@@ -254,15 +257,16 @@ pub fn take_until_size(buf: &[u8], byte_size: u32) -> IResult<&[u8], &[u8]> {
...
+fn parse_meta_flag_values_as_slice(input: &[u8]) -> IResult<&[u8], Vec<(u8, Option<&[u8]>)>> {
     if input.is_empty() || input == b"\r\n" {
         Ok((input, Vec::new()))
     } else {
         many0(map(
             tuple((
                 space1,
                 map(take(1usize), |s: &[u8]| s[0]),
-                take_while1(|c: u8| c != b'\r' && c != b' '), // finding a space means more flags, finding \r means end of flags
+                opt(take_while1(|c: u8| c != b'\r' && c != b' ')), // finding a space means more flags, finding \r means end of flags
```
Comment (mrattle) [inline src/parser/meta.rs:269]: Needed to add this in when encountering single character tokens like X, W, Z which never have anything following them.  This was leading to nom errors in any case that they showed up.

File: src/parser/meta.rs [hunk 2]
```diff
@@ -375,7 +379,7 @@ fn map_meta_flag(
 }
 
 fn construct_value_from_meta_values(
-    meta_values_array: Vec<(u8, &[u8])>,
+    meta_values_array: Vec<(u8, Option<&[u8]>)>,
```
Comment (mrattle) [inline src/parser/meta.rs:382]: Consequence of the parsing change, the tokens that correspond to the first char of each flag are now optional

File: src/lib.rs [hunk 3]
```diff
@@ -308,6 +308,70 @@ impl Client {
...
+        self.conn.write_all(b"\r\n").await?;
+        self.conn.flush().await?;
+
+        match self.drive_receive(parse_meta_set_response).await? {
+            Response::Status(Status::Stored) => Ok(None),
+            Response::Status(s) => Err(s.into()),
+            Response::Data(d) => d
+                .map(|mut items| {
+                    if items.len() != 1 {
+                        Err(Status::Error(ErrorKind::Protocol(None)).into())
```
Comment (grcooper) [inline src/lib.rs:363]: Similar to the meta get PR, let's make a more verbose error here.

**Review by grcooper — APPROVED** [review]
great set of tests, lgtm with one nit.

File: src/lib.rs [hunk 4]
```diff
@@ -308,6 +308,63 @@ impl Client {
...
+        self.conn.write_all(vr.as_ref()).await?;
+        self.conn.write_all(b"\r\n").await?;
+        self.conn.flush().await?;
+
+        match self.drive_receive(parse_meta_set_response).await? {
+            Response::Status(Status::Stored) => Ok(None),
+            Response::Status(s) => Err(s.into()),
+            Response::Data(d) => d
+                .map(|mut items| {
+                    let item = items.remove(0);
```
Comment (nickamorim) [inline src/lib.rs:360]: Potentially a dumb question but could this ever throw an index out of bounds?

**Review by nickamorim — APPROVED** [review]
